[WPE] WPE Platform: implement WebPasteboardProxy::readURLFromPasteboard

carlosgcampos opened the pull request:
#### 2df4e5d4891197f9662680b7b16b70672f230233
<pre>
[WPE] WPE Platform: implement WebPasteboardProxy::readURLFromPasteboard
<a href="https://bugs.webkit.org/show_bug.cgi?id=292927">https://bugs.webkit.org/show_bug.cgi?id=292927</a>

Reviewed by Alejandro G. Castro.

Also make sure the strstegy methods are called in
Pasteboard::allPasteboardItemInfo and Pasteboard::pasteboardItemInfo for
WPE port too.

Canonical link: <a href="https://commits.webkit.org/294953@main">https://commits.webkit.org/294953@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a83c1142df0e23ceb58447d9e95c8dc4e74a18d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103609 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23291 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13612 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108785 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54256 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105649 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23647 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31842 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78732 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106615 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18345 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93468 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59067 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/11515 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53610 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87939 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11574 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111163 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30756 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22680 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87727 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31117 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89668 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87374 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32251 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9972 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25110 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16812 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30684 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/35996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30484 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33815 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32045 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alexgcastro commented on Source/WebKit/UIProcess/wpe/WebPasteboardProxyWPE.cpp:
```diff
@@ -292,6 +292,21 @@ void WebPasteboardProxy::getPasteboardItemsCount(IPC::Connection&, const String&
+        auto* clipboard = wpe_display_get_clipboard(wpe_display_get_primary());
+        if (GRefPtr<GBytes> bytes = adoptGRef(wpe_clipboard_read_bytes(clipboard, "text/uri-list"))) {
+            auto buffer = FragmentedSharedBuffer::create(bytes.get())->makeContiguous();
+            completionHandler(String(buffer->span()), { });
```
Is it possible that buffer could be null?

carlosgcampos replied:
No, it can be empty but not null, both makeContiguous() and create() return a Ref

alexgcastro commented on Source/WebKit/UIProcess/wpe/WebPasteboardProxyWPE.cpp:
```diff
@@ -292,6 +292,21 @@ void WebPasteboardProxy::getPasteboardItemsCount(IPC::Connection&, const String&
 {
+#if ENABLE(WPE_PLATFORM)
+    if (usingWPEPlatformAPI()) {
+        if (index) {
```
Can you document what does it mean that index is not zero?

carlosgcampos replied:
This is because we don't support multiple items in the clipboard

alexgcastro reviewed (approved):
LGTM!

carlosgcampos committed:
[WPE] WPE Platform: implement WebPasteboardProxy::readURLFromPasteboard
https://bugs.webkit.org/show_bug.cgi?id=292927

Reviewed by Alejandro G. Castro.

Also make sure the strstegy methods are called in
Pasteboard::allPasteboardItemInfo and Pasteboard::pasteboardItemInfo for
WPE port too.

Canonical link: https://commits.webkit.org/294953@main
